feat(mcp): tool title

AVtheking opened the pull request:
## Background
MCP supports for human readable title for the tool which is currently not supported by ai sdk ToolUIPart.

## Summary

Add a way to be able to use the mcp tool title in ui.

## Manual Verification

Wrote an next js eg with mcp server and tools with title and using that title in the ui as display name for the tool.

## Checklist

- [x] Tests have been added / updated (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)
- [x] A _patch_ changeset for relevant packages has been added (for bug fixes / features - run `pnpm changeset` in the project root)
- [x] Formatting issues have been fixed (run `pnpm prettier-fix` in the project root)
- [x] I have reviewed this pull request (self-review)


## Related Issues
Fixes #9163

Commit:
title for the tool

Commit:
changeset

Commit:
prettier fix

Commit:
update lock file

AVtheking committed:
Merge branch 'main' into feat/tool-title

Commit:
updates tests

Commit:
updates lock file

Commit:
fix tests

AVtheking committed:
prettier

AVtheking committed:
prettier

AVtheking committed:
prettier

AVtheking committed:
fixed svelte test

AVtheking commented:
Hey @gr2m, I have tested this and this working correctly, but might have missed some cases where title needs to be extracted. Could you PTAL, I will resolve the conflicts then.

aayush-kapoor commented on examples/next/app/api/mcp-tool-title/route.ts:
same as above - add tool title to an existing example

aayush-kapoor commented on packages/ai/src/generate-text/__snapshots__/stream-text.test.ts.snap:
it's not a good idea to have the title be 'undefined' in the tests

aayush-kapoor commented on packages/ai/src/generate-text/stream-text.test.ts:
change undefined

aayush-kapoor commented:
@AVtheking please address the comments above

AVtheking commented:
> @AVtheking please address the comments above

Thanks for the review, I would address these and get back.

aayush-kapoor commented on packages/ai/src/generate-text/parse-tool-call.test.ts:
```diff
@@ -419,10 +429,93 @@ describe('parseToolCall', () => {
+        tools: {
+          weather: {
+            type: 'dynamic',
+            title: '🌤️ Weather Information',
```
please remove the emojis from all the tests everywhere.

AVtheking commented on packages/ai/src/generate-text/run-tools-transformation.test.ts:
Moreover these tool titles are for MCP tools right now so I think this make sense to be undefined for other methods right ? as this field is clearly optional

AVtheking committed:
update eg and test

AVtheking committed:
Merge branch 'main' into feat/tool-title

AVtheking committed:
update test

AVtheking committed:
fix test

lgrammel commented on packages/ai/src/ui-message-stream/ui-message-chunks.ts:
```diff
@@ -41,11 +41,13 @@ export const uiMessageChunkSchema = lazySchema(() =>
         type: z.literal('tool-input-delta'),
         toolCallId: z.string(),
         inputTextDelta: z.string(),
+        title: z.string().optional(),
```
we should not send this on every delta, it can be a lot of data

lgrammel commented:
I feel somewhat unesasy about this PR, because it sends a static tool property for the UI many times in the same ui message stream. Next someone will do something similar for mcp tool descriptions, and then we will see complaints about slow ui message stream performance. The core reason is mixing a property that is scoped on the tool level into the too call level (which has some DX advantages). Can you minimize how often the tool title over the wire? e.g. it does not need to be part of the tool call delta or tool input end events.

AVtheking commented:
> I feel somewhat unesasy about this PR, because it sends a static tool property for the UI many times in the same ui message stream. Next someone will do something similar for mcp tool descriptions, and then we will see complaints about slow ui message stream performance. The core reason is mixing a property that is scoped on the tool level into the too call level (which has some DX advantages). Can you minimize how often the tool title over the wire? e.g. it does not need to be part of the tool call delta or tool input end events.

Got it, make sense will do

AVtheking committed:
remove title form other events

AVtheking committed:
Merge branch 'main' into feat/tool-title

AVtheking committed:
Merge branch 'main' into feat/tool-title

Commit:
Version Packages (beta) (#9760)

This PR was opened by the [Changesets
release](https://github.com/changesets/action) GitHub action. When
you're ready to do a release, you can merge this and the packages will
be published to npm automatically. If you're not ready to do a release
yet, that's fine, whenever you add more changesets to main, this PR will
be updated.

⚠️⚠️⚠️⚠️⚠️⚠️

`main` is currently in **pre mode** so this branch has prereleases
rather than normal releases. If you want to exit prereleases, run
`changeset pre exit` on `main`.

⚠️⚠️⚠️⚠️⚠️⚠️

# Releases
## @ai-sdk/amazon-bedrock@4.0.0-beta.41

### Patch Changes

-   Updated dependencies [7c4328e]
    -   @ai-sdk/anthropic@3.0.0-beta.37

## @ai-sdk/anthropic@3.0.0-beta.37

### Patch Changes

- 7c4328e: Adds url-based pdf and image support for anthropic tool
results

## @ai-sdk/google-vertex@4.0.0-beta.45

### Patch Changes

-   Updated dependencies [7c4328e]
    -   @ai-sdk/anthropic@3.0.0-beta.37

Co-authored-by: vercel-ai-sdk[bot] <225926702+vercel-ai-sdk[bot]@users.noreply.github.com>

lgrammel committed:
chore(ai): add unit tests and jsdoc for Output (#9768)

## Background

`Output.text()` and `Output.object()` had only limited unit tests and no
jsdoc.

## Summary

add unit tests and jsdoc for Output

## Future Work

* add more Output options

lgrammel committed:
feat(ai): add Output.array() (#9769)

## Background

`experimental_output` on `generateText` and `streamText` should support
the same options as `generateObject` and eventually replace it.

## Summary

* Add `Output.array` option
* Add output response format tests

## Manual Verification

- [x] run `examples/ai-core/src/generate-text/openai-output-array.ts`
- [x] run `examples/ai-core/src/stream-text/openai-output-array.ts`

## Future Work

* custom array name
* array element stream
* enum type
* json type

Commit:
Version Packages (beta) (#9773)

This PR was opened by the [Changesets
release](https://github.com/changesets/action) GitHub action. When
you're ready to do a release, you can merge this and the packages will
be published to npm automatically. If you're not ready to do a release
yet, that's fine, whenever you add more changesets to main, this PR will
be updated.

⚠️⚠️⚠️⚠️⚠️⚠️

`main` is currently in **pre mode** so this branch has prereleases
rather than normal releases. If you want to exit prereleases, run
`changeset pre exit` on `main`.

⚠️⚠️⚠️⚠️⚠️⚠️

-   d5b25ee: feat(ai): add Output.array()

-   Updated dependencies [d5b25ee]
    -   ai@6.0.0-beta.69

-   Updated dependencies [d5b25ee]
    -   ai@6.0.0-beta.69

-   Updated dependencies [d5b25ee]
    -   ai@6.0.0-beta.69

-   Updated dependencies [d5b25ee]
    -   ai@6.0.0-beta.69

-   Updated dependencies [d5b25ee]
    -   ai@6.0.0-beta.69

-   Updated dependencies [d5b25ee]
    -   ai@6.0.0-beta.69

-   Updated dependencies [d5b25ee]
    -   ai@6.0.0-beta.69

Co-authored-by: vercel-ai-sdk[bot] <225926702+vercel-ai-sdk[bot]@users.noreply.github.com>

lgrammel committed:
feat(ai): add Output.choice() (#9772)

## Background

`experimental_output` on `generateText` and `streamText` should support
the same options as `generateObject` and eventually replace it.

## Summary

* Add `Output.choice` option

## Manual Verification

- [x] run `examples/ai-core/src/generate-text/openai-output-choice.ts`
- [x] run `examples/ai-core/src/stream-text/openai-output-choice.ts`

Commit:
Version Packages (beta) (#9776)

This PR was opened by the [Changesets
release](https://github.com/changesets/action) GitHub action. When
you're ready to do a release, you can merge this and the packages will
be published to npm automatically. If you're not ready to do a release
yet, that's fine, whenever you add more changesets to main, this PR will
be updated.

⚠️⚠️⚠️⚠️⚠️⚠️

`main` is currently in **pre mode** so this branch has prereleases
rather than normal releases. If you want to exit prereleases, run
`changeset pre exit` on `main`.

⚠️⚠️⚠️⚠️⚠️⚠️

# Releases
## ai@6.0.0-beta.70

### Patch Changes

-   d7bae86: feat(ai): add Output.choice()

## @ai-sdk/angular@2.0.0-beta.70

### Patch Changes

-   Updated dependencies [d7bae86]
    -   ai@6.0.0-beta.70

## @ai-sdk/langchain@2.0.0-beta.70

### Patch Changes

-   Updated dependencies [d7bae86]
    -   ai@6.0.0-beta.70

## @ai-sdk/llamaindex@2.0.0-beta.70

### Patch Changes

-   Updated dependencies [d7bae86]
    -   ai@6.0.0-beta.70

## @ai-sdk/react@3.0.0-beta.70

### Patch Changes

-   Updated dependencies [d7bae86]
    -   ai@6.0.0-beta.70

## @ai-sdk/rsc@2.0.0-beta.70

### Patch Changes

-   Updated dependencies [d7bae86]
    -   ai@6.0.0-beta.70

## @ai-sdk/svelte@4.0.0-beta.70

### Patch Changes

-   Updated dependencies [d7bae86]
    -   ai@6.0.0-beta.70

## @ai-sdk/vue@3.0.0-beta.70

### Patch Changes

-   Updated dependencies [d7bae86]
    -   ai@6.0.0-beta.70

Co-authored-by: vercel-ai-sdk[bot] <225926702+vercel-ai-sdk[bot]@users.noreply.github.com>

lgrammel committed:
feat(ai): stable structured output on generateText, streamText, and ToolLoopAgent (#9778)

## Background

The experimental structured outputs API for streamText and generateText
has been around for quite some time. In AI SDK 6, we are moving it to
stable and adding features to achieve parity with generateObject and
streamObject to support the Agent abstraction.

## Summary

Move structured outputs for streamText and generateText to stable.

## Future Work

* add json mode
* add array element stream

Commit:
Version Packages (beta) (#9779)

This PR was opened by the [Changesets
release](https://github.com/changesets/action) GitHub action. When
you're ready to do a release, you can merge this and the packages will
be published to npm automatically. If you're not ready to do a release
yet, that's fine, whenever you add more changesets to main, this PR will
be updated.

⚠️⚠️⚠️⚠️⚠️⚠️

`main` is currently in **pre mode** so this branch has prereleases
rather than normal releases. If you want to exit prereleases, run
`changeset pre exit` on `main`.

⚠️⚠️⚠️⚠️⚠️⚠️

# Releases
## ai@6.0.0-beta.71

### Patch Changes

- 077aea3: feat(ai): stable structured output on generateText,
streamText, and ToolLoopAgent

## @ai-sdk/angular@2.0.0-beta.71

### Patch Changes

-   Updated dependencies [077aea3]
    -   ai@6.0.0-beta.71

## @ai-sdk/langchain@2.0.0-beta.71

### Patch Changes

-   Updated dependencies [077aea3]
    -   ai@6.0.0-beta.71

## @ai-sdk/llamaindex@2.0.0-beta.71

### Patch Changes

-   Updated dependencies [077aea3]
    -   ai@6.0.0-beta.71

## @ai-sdk/react@3.0.0-beta.71

### Patch Changes

-   Updated dependencies [077aea3]
    -   ai@6.0.0-beta.71

## @ai-sdk/rsc@2.0.0-beta.71

### Patch Changes

-   Updated dependencies [077aea3]
    -   ai@6.0.0-beta.71

## @ai-sdk/svelte@4.0.0-beta.71

### Patch Changes

-   Updated dependencies [077aea3]
    -   ai@6.0.0-beta.71

## @ai-sdk/vue@3.0.0-beta.71

### Patch Changes

-   Updated dependencies [077aea3]
    -   ai@6.0.0-beta.71

Co-authored-by: vercel-ai-sdk[bot] <225926702+vercel-ai-sdk[bot]@users.noreply.github.com>

gr2m committed:
ci(triage): add another example (#9729)

aayush-kapoor committed:
feat(ai): add OAuth for MCP clients + refactor to new package (#9127)

## Background

Refactor mcp into it's own separate package + add OAuth for MCP clients

## Summary

Created new package `@ai-sdk/mcp`, added OAuth provider

## Manual Verification

Manual tests added, e2e example added to verify oauth works in
`example/mcp/mcp-with-auth/client.ts`

## Checklist

- [x] Tests have been added / updated (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] A _patch_ changeset for relevant packages has been added (for bug
fixes / features - run `pnpm changeset` in the project root)
- [x] Formatting issues have been fixed (run `pnpm prettier-fix` in the
project root)

## Future Work

- Continue refactoring, add docs
- log warning or throw error if wrong transport protocol is set for an
mcp server

## Related Issues

towards #8717, #9337 and #6717

---------

Co-authored-by: Gregor Martynus <39992+gr2m@users.noreply.github.com>

Commit:
Version Packages (beta) (#9783)

This PR was opened by the [Changesets
release](https://github.com/changesets/action) GitHub action. When
you're ready to do a release, you can merge this and the packages will
be published to npm automatically. If you're not ready to do a release
yet, that's fine, whenever you add more changesets to main, this PR will
be updated.

⚠️⚠️⚠️⚠️⚠️⚠️

`main` is currently in **pre mode** so this branch has prereleases
rather than normal releases. If you want to exit prereleases, run
`changeset pre exit` on `main`.

⚠️⚠️⚠️⚠️⚠️⚠️

# Releases
## @ai-sdk/mcp@1.0.0-beta.0

### Major Changes

- eca63f3: feat(ai): add OAuth for MCP clients + refactor to new package

    This change replaces

    ```ts
    import { experimental_createMCPClient } from 'ai';
    import { Experimental_StdioMCPTransport } from 'ai/mcp-stdio';
    ```

    with

    ```ts
    import { experimental_createMCPClient } from '@ai-sdk/mcp';
import { Experimental_StdioMCPTransport } from '@ai-sdk/mcp/mcp-stdio';
    ```

## ai@6.0.0-beta.72

### Patch Changes

- eca63f3: feat(ai): add OAuth for MCP clients + refactor to new package

    This change replaces

    ```ts
    import { experimental_createMCPClient } from 'ai';
    import { Experimental_StdioMCPTransport } from 'ai/mcp-stdio';
    ```

    with

    ```ts
    import { experimental_createMCPClient } from '@ai-sdk/mcp';
import { Experimental_StdioMCPTransport } from '@ai-sdk/mcp/mcp-stdio';
    ```

## @ai-sdk/angular@2.0.0-beta.72

### Patch Changes

-   Updated dependencies [eca63f3]
    -   ai@6.0.0-beta.72

## @ai-sdk/langchain@2.0.0-beta.72

### Patch Changes

-   Updated dependencies [eca63f3]
    -   ai@6.0.0-beta.72

## @ai-sdk/llamaindex@2.0.0-beta.72

### Patch Changes

-   Updated dependencies [eca63f3]
    -   ai@6.0.0-beta.72

## @ai-sdk/react@3.0.0-beta.72

### Patch Changes

-   Updated dependencies [eca63f3]
    -   ai@6.0.0-beta.72

## @ai-sdk/rsc@2.0.0-beta.72

### Patch Changes

-   Updated dependencies [eca63f3]
    -   ai@6.0.0-beta.72

## @ai-sdk/svelte@4.0.0-beta.72

### Patch Changes

-   Updated dependencies [eca63f3]
    -   ai@6.0.0-beta.72

## @ai-sdk/vue@3.0.0-beta.72

### Patch Changes

-   Updated dependencies [eca63f3]
    -   ai@6.0.0-beta.72

Co-authored-by: vercel-ai-sdk[bot] <225926702+vercel-ai-sdk[bot]@users.noreply.github.com>

AVtheking committed:
Merge branch 'main' into feat/tool-title

AVtheking commented:
Strange how did this added commits while merging main

lgrammel commented on packages/mcp/src/tool/types.ts:
```diff
@@ -103,6 +103,17 @@ const ToolSchema = z
+          destructiveHint: z.optional(z.boolean()),
+          idempotentHint: z.optional(z.boolean()),
+          openWorldHint: z.optional(z.boolean()),
+        })
```
limit to title

AVtheking replied:
just added all the fields supported in annotations.

lgrammel commented:
@AVtheking some last comment but looks 98% there. Great job!

AVtheking committed:
nits
